File: leetcode/lc-us/cpp/2132.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    bool possibleToStamp(vector<vector<int>> &grid, int h, int w) {
        int n = (int) grid.size(), m = (int) grid[0].size();
        int s[n + 1][m + 1], diff[n + 2][m + 2];
        memset(s, 0, sizeof s), memset(diff, 0, sizeof diff);
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= m; j++) {
                s[i][j] = s[i - 1][j] + s[i][j - 1] - s[i - 1][j - 1] + grid[i - 1][j - 1];
            }
        }
        for (int i = 1; i + h - 1 <= n; i++) {
            for (int j = 1; j + w - 1 <= m; j++) {
                int cnt = s[i + h - 1][j + w - 1] - s[i + h - 1][j - 1] - s[i - 1][j + w - 1] + s[i - 1][j - 1];
                if (!cnt) {
                    diff[i][j] += 1;
                    diff[i][j + w] -= 1;
                    diff[i + h][j] -= 1;
                    diff[i + h][j + w] += 1;
                }
            }
        }
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= m; j++) {
                diff[i][j] += diff[i - 1][j] + diff[i][j - 1] - diff[i - 1][j - 1];
                if (!grid[i - 1][j - 1]) {
                    if (!diff[i][j]) return false;
                }
            }
        }
        return true;
    }
};

File: leetcode/lc-us/py/2132.py
from typing import List


class Solution:
    def possibleToStamp(self, grid: List[List[int]], h: int, w: int) -> bool:
        n, m = len(grid), len(grid[0])
        s = [[0] * (m + 1) for _ in range(n + 1)]
        diff = [[0] * (m + 2) for _ in range(n + 2)]
        for i in range(1, n + 1):
            for j in range(1, m + 1):
                s[i][j] = s[i - 1][j] + s[i][j - 1] - s[i - 1][j - 1] + grid[i - 1][j - 1]
        for i in range(1, n - h + 2):
            for j in range(1, m - w + 2):
                if s[i + h - 1][j + w - 1] - s[i + h - 1][j - 1] - s[i - 1][j + w - 1] + s[i - 1][j - 1] == 0:
                    diff[i][j] += 1
                    diff[i + h][j] -= 1
                    diff[i][j + w] -= 1
                    diff[i + h][j + w] += 1

        for i in range(1, n + 1):
            for j in range(1, m + 1):
                diff[i][j] += diff[i - 1][j] + diff[i][j - 1] - diff[i - 1][j - 1]
                if grid[i - 1][j - 1] == 0 and diff[i][j] == 0:
                    return False
        return True
